### DISCUSSION TITLE
How do I configure haskell-tools.nvim to tell haskell-language-server that my format provider is fourmolu?

### DISCUSSION THREAD

**Discussion by juarpr** [mrcjkb/haskell-tools.nvim#405] [Q&A] [answered]
Hi everyone, How do I configure haskell-tools.nvim to tell haskell-language-server that my format provider is fourmolu?
If you could do it through haskell-tool.nvim instead of haskell-language-server directly, the integration with neovim would be better

**Comment by mrcjkb** [comment, marked as answer]
you can configure haskell-language-server by passing a Lua table via the haskell-tools config. 

See `:h haskell-tools.config` and `:h haskell-tools.lsp.ClientOpts`.

**Reply by juarpr** [reply]
That's nice! Thank you!

**Reply by juarpr** [reply]
I know that fourmolu is the default formatter.  I'm getting [LSP][haskell-tools.nvim] timeout. How I would write my on_attach = function(client, bufnr, ht) in vim.g.haskell_tools? I want to configure haskell-tools.nvim to increase the timeout limits.

**Reply by mrcjkb** [reply]
If you're getting a timeout error when formatting, that's nothing to do with your haskell-tools configuration. See `:h vim.lsp.buf.format()`.

**Reply by juarpr** [reply]
I have written in `~/.config/nvim/after/ftplugin/haskell.lua`:
```lua
vim.keymap.set('n', '<space>f', function()
  -- With the async=false flag, the format request is guaranteed to be blocked until it is complete,
  -- so that it finishes formatting completely before emptying the file to disk.
  vim.lsp.buf.format { async = false }
end, opts)
```
Some buffer formats and others it doesn't, it has to do with the size of the haskell module. Fourmolu from the command line performs formatting whatever the file size. 
I don't know how to adapt this code to solve the integration of fourmolu to neovim, (I wrote it in haskell.lua):
``` lua
-- 1
vim.api.nvim_create_autocmd("LspAttach", {
  group = vim.api.nvim_create_augroup("lsp", { clear = true }),
  callback = function(args)
    -- 2
    vim.api.nvim_create_autocmd("BufWritePre", {
      -- 3
      buffer = args.buf,
      callback = function()
        -- 4 + 5
        vim.lsp.buf.format {async = false, id = args.data.client_id }
      end,
    })
  end
})
```

**Reply by mrcjkb** [reply]
You probably don't want to create an `LspAttach` autocommand in `ftplugin/haskell.lua`.
It might get created after the LSP client attaches and thus never triggered.

**Reply by juarpr** [reply]
The `:LspInfo` command gets:
```
==============================================================================
lspconfig: require("lspconfig.health").check()

LSP configs active in this session (globally) ~
- Configured servers: lua_ls
- OK Deprecated servers: (none)

LSP configs active in this buffer (id=1) ~
- Language client log: ~/.local/state/nvim-fourm/lsp.log
- Detected filetype: `haskell`
- 1 client(s) attached to this buffer
- Client: haskell-tools.nvim (id: 1, bufnr: [1])
  filetypes:       haskell, lhaskell
  root directory:  ~/calpolifourm
  cmd:             ~/calpolifourm/haskell-language-server-wrapper --lsp --logfile /tmp/nvim-fourm.beatriz/NDDqUi/0-haskell-language-server.log
  autostart:       false

Docs for active configs: ~
- 
```
I don't know how to tell my autocommand that my LSP client is haskell-tools.nvim. Even when I type autocommand into init.lua I still get:
```
   Warn  06:37:04 notify.warn [LSP][haskell-tools.nvim] timeout
   Warn  06:37:05 notify.warn [LSP][haskell-tools.nvim] timeout
06:37:05 msg_show "~/calpolifourm/app/Main.hs" 
06:37:05 msg_show "~/calpolifourm/app/Main.hs" 1315L, 63275B
```

**Reply by mrcjkb** [reply]
I don't know what you're trying to do.
If you call `vim.lsp.buf.format()` with `sync = true`, it can time out, as is described in the `:h vim.lsp.buf.format()` documentation.
This is not specific to haskell-tools or haskell-language-server.

**Reply by juarpr** [reply]
I fixed it, thanks for the patience:
``` lua
vim.keymap.set('n', '<space>f', function()
  -- With the async=false flag, the format request is guaranteed to be blocked until it is complete,
  -- so that it finishes formatting completely before emptying the file to disk.
  vim.lsp.buf.format { async = false, timeout_ms = 10000 }
end, opts)
```
With timeout_ms=10000, fourmolu formats it and does not issue any timeout warning

**Reply by juarpr** [reply]
With the autocommand, it formats it when saving, but issues the warning:
``` lua
-- 1
vim.api.nvim_create_autocmd("LspAttach", {
  group = vim.api.nvim_create_augroup("lsp", { clear = true }),
  callback = function(args)
    -- 2
    vim.api.nvim_create_autocmd("BufWritePre", {
      -- 3
      buffer = args.buf,
      callback = function()
        -- 4 + 5
        vim.lsp.buf.format { async = false, timeout_ms = 10000, id = args.data.client_id }
      end,
    })
  end
})
```
I am satisfied, even when I receive the timeout warning. Congratulations on your great haskell-tools.nvim plugin my goal is for it to take care of everything related to haskell-language-server.
